File: packages/core/src/teams.ts
import { buildBoard, type Board } from "./board";
import * as conversations from "./conversations";
import type { Context } from "./conversations";
import { TransportError } from "./errors";
import type { Payload, SealedFile } from "./payload";
import { moderates, refreshRoster, type RosterEntry } from "./roster";
import type { StoredConversation, TeamRole } from "./store";

/**
 * Teams: a private board of posts over an ordinary MLS conversation.
 *
 * Built on `conversations.ts`, never beside it. A team is a conversation with
 * `kind = 'team'`: the same group, the same envelopes, the same sync, the same
 * outbox. Adding somebody is `addTo`; a post is `sendPayload`; reactions, edits
 * and take-backs are the ordinary ones, because posts and comments are rows in
 * `messages` like any message. What lives here is only what a team has and a
 * group does not -- roles, a board, and the few server routes that manage them
 * (`apps/server/src/teams.rs`).
 *
 * Everything a member writes is inside the ciphertext. The server learns that
 * the team exists, who is in it with which role, and when envelopes flow --
 * not its name, not its description, not a word of any post.
 */

/** A team as the server lists it for its member. Nothing about its content. */
export interface TeamSummary {
  conversation_id: string;
  /** The caller's own role. */
  role: TeamRole;
  member_count: number;
  created_at_ms: number;
}

/** The most people a team may hold. `MAX_MEMBERS` in `apps/server/src/teams.rs`. */
export const MAX_MEMBERS = 200;

/** Every team this account is in, from the server. */
export function listTeams(ctx: Context): Promise<TeamSummary[]> {
  return ctx.transport.getAuth<TeamSummary[]>("/v1/teams");
}

/** Who is in a team and with which role, read fresh and remembered beside it. */
export function roster(ctx: Context, teamId: string): Promise<RosterEntry[]> {
  return refreshRoster(ctx.transport, ctx.store, teamId);
}

/** This account's role in a team, from the roster last read. */
export async function myRole(ctx: Context, teamId: string): Promise<TeamRole | undefined> {
  const me = (await ctx.store.account())?.handle;
  const team = await ctx.store.conversation(teamId);
  return me === undefined ? undefined : team?.roles?.[me];
}

// ------------------------------------------------------------------- creating

/**
 * Starts a team with this account as its owner and only member.
 *
 * The name is kept on this device and **not sent**: there is nobody yet to
 * send it to, and earlier ciphertext cannot be read by a later member anyway.
 * It goes out with the description and picture after every add
 * (`addPeople`), which is how somebody new learns what the team is called.
 */
export async function createTeam(ctx: Context, name: string, description = ""): Promise<string> {
  const title = name.trim();
  if (title === "") throw new TransportError("rejected", "A team needs a name.");
  const now = (ctx.now ?? Date.now)();
  const id = (ctx.uuid ?? (() => globalThis.crypto.randomUUID()))();

  ctx.crypto.createGroup(ctx.device, id, now);
  await ctx.store.setMlsState(ctx.device.exportState());
  const created = await ctx.transport.postAuth<TeamSummary>("/v1/teams", { conversation_id: id });
  if (created.conversation_id !== id) {
    throw new TransportError("rejected", "The server returned a different team id.");
  }

  const me = (await ctx.store.account())?.handle;
  const team: StoredConversation = {
    id,
    title,
    kind: "team",
    epoch: 0,
    syncedTo: 0,
    lastMessage: null,
    updatedAtMs: now,
  };
  if (me !== undefined) {
    team.members = [me];
    team.roles = { [me]: "owner" };
  }
  if (description.trim() !== "") team.description = description.trim();
  await ctx.store.putConversation(team);
  return id;
}

// ------------------------------------------------------------------ membership

/** Why somebody could not be added, in words the add-people dialog can show. */
export interface AddOutcome {
  handle: string;
  added: boolean;
  /** Present when `added` is false: what the server or the device said. */
  reason?: string;
}

/**
 * Adds people, one commit each, and then tells the team what it is called.
 *
 * One commit per person because that is what `addTo` is, and a partial
 * success is honest here: the people who could be added are, and each one who
 * could not says why. The server decides who may be added -- only an owner or
 * admin, not somebody blocked or private without an invitation, not past the
 * cap -- and answers the same way it answers a group, so a refusal never says
 * *which* of blocked or private it was.
 *
 * After the adds, the name, the description and the picture are sent again.
 * A new member reads nothing sent before their Welcome, so without this they
 * would see "New team" until somebody next renamed it.
 */
export async function addPeople(ctx: Context, teamId: string, handles: string[]): Promise<AddOutcome[]> {
  const outcomes: AddOutcome[] = [];
  for (const handle of handles) {
    try {
      await conversations.addTo(ctx, teamId, handle);
      outcomes.push({ handle, added: true });
    } catch (error) {
      outcomes.push({
        handle,
        added: false,
        reason: error instanceof Error ? error.message : "That could not be done.",
      });
    }
  }
  if (outcomes.some((outcome) => outcome.added)) {
    await announce(ctx, teamId);
    await roster(ctx, teamId);
  }
  return outcomes;
}

/** Sends what a new member needs to name and describe the team. */
async function announce(ctx: Context, teamId: string): Promise<void> {
  const team = await ctx.store.conversation(teamId);
  if (!team) return;
  if (team.title) await conversations.rename(ctx, teamId, team.title);
  if (team.description) {
    await conversations.sendPayload(ctx, teamId, { kind: "team_meta", description: team.description });
  }
  if (team.avatar) {
    await conversations.sendPayload(ctx, teamId, JSON.parse(team.avatar) as Payload);
  }
}

/** Takes somebody out: the server's rule first, then the commit that locks them out. */
export async function removePerson(ctx: Context, teamId: string, handle: string): Promise<void> {
  await conversations.removeFrom(ctx, teamId, handle);
  await roster(ctx, teamId);
}

/** Makes somebody an admin, or an admin a member again. */
export async function setRole(
  ctx: Context,
  teamId: string,
  handle: string,
  role: Exclude<TeamRole, "owner">,
): Promise<void> {
  await ctx.transport.patchAuth<void>(
    `/v1/teams/${teamId}/members/${encodeURIComponent(handle)}`,
    { role },
  );
  await roster(ctx, teamId);
}

/** Hands the team to somebody already in it. This account stays, as an admin. */
export async function transfer(ctx: Context, teamId: string, handle: string): Promise<void> {
  await ctx.transport.postAuth<void>(`/v1/teams/${teamId}/transfer`, { handle });
  await roster(ctx, teamId);
}

/**
 * Leaves a team. Anybody but the owner.
 *
 * The routing row goes and this device forgets the team. Its MLS leaf stays
 * until an owner's or admin's device commits it out (`reconcile`) -- a member
 * cannot commit their own removal -- and until then the server no longer
 * hands this device anything to read with it.
 */
export async function leave(ctx: Context, teamId: string): Promise<void> {
  await ctx.transport.postAuth<void>(`/v1/teams/${teamId}/leave`, {});
  await ctx.store.forgetConversation(teamId);
}

/** Deletes a team for everybody. The owner only; the server refuses anybody else. */
export async function deleteTeam(ctx: Context, teamId: string): Promise<void> {
  await ctx.transport.deleteAuth(`/v1/teams/${teamId}`);
  await ctx.store.forgetConversation(teamId);
}

/**
 * Commits out every device that is still in the MLS group but no longer on
 * the server's list -- somebody who left, or whose routing removal landed and
 * whose commit did not.
 *
 * Only an owner's or admin's device does this, and only against a list read
 * just now: a stale list would name a member added since as somebody to
 * remove. Returns how many devices it removed.
 */
export async function reconcile(ctx: Context, teamId: string): Promise<number> {
  await conversations.discover(ctx);
  await roster(ctx, teamId);
  if (!moderates(await myRole(ctx, teamId))) return 0;

  const team = await ctx.store.conversation(teamId);
  const group = ctx.crypto.loadGroup(ctx.device, teamId, (ctx.now ?? Date.now)());
  if (!team?.memberDevices || !group) return 0;
  const me = (await ctx.store.identity())?.deviceId;
  const listed = new Set(Object.keys(team.memberDevices));
  const gone = group
    .members()
    .map((member) => member.deviceId)
    .filter((deviceId) => deviceId !== me && !listed.has(deviceId));
  for (const deviceId of gone) await conversations.removeDevice(ctx, teamId, deviceId);
  return gone.length;
}

// -------------------------------------------------------------------- the board

/** What a new post may carry. */
export interface PostDraft {
  title?: string;
  body: string;
  files?: SealedFile[];
}

/**
 * Posts to the board. Answers the post's name, which comments, reactions and
 * pins point at.
 *
 * `null` from `sendPayload` means the post is queued offline; the name is
 * still the one it will have.
 */
export async function post(ctx: Context, teamId: string, draft: PostDraft): Promise<string> {
  const body = draft.body.trim();
  const files = draft.files ?? [];
  if (body === "" && files.length === 0) {
    throw new TransportError("rejected", "Write something, or attach a file.");
  }
  const id = (ctx.uuid ?? (() => globalThis.crypto.randomUUID()))();
  const payload: Payload = { kind: "team_post", id, body };
  const title = draft.title?.trim();
  if (title) payload.title = title;
  if (files.length > 0) payload.files = files;
  await conversations.sendPayload(ctx, teamId, payload);
  return id;
}

/** Comments on a post, or answers a top-level comment on it. */
export async function comment(
  ctx: Context,
  teamId: string,
  postId: string,
  body: string,
  parent?: string,
): Promise<string> {
  const text = body.trim();
  if (text === "") throw new TransportError("rejected", "A comment needs some words.");
  const id = (ctx.uuid ?? (() => globalThis.crypto.randomUUID()))();
  const payload: Payload = { kind: "team_comment", id, post: postId, body: text };
  if (parent !== undefined) payload.parent = parent;
  await conversations.sendPayload(ctx, teamId, payload);
  return id;
}

/**
 * Pins a post, or unpins it. Offered to owners and admins only, and refused
 * here for anybody else -- the other devices would ignore it anyway, and a pin
 * that shows on one screen and nowhere else is worse than none.
 */
export async function pin(ctx: Context, teamId: string, postId: string, pinned: boolean): Promise<void> {
  await requireModerator(ctx, teamId);
  await conversations.sendPayload(ctx, teamId, { kind: "team_pin", post: postId, pinned });
}

/**
 * Asks every member's device to stop showing a post or comment.
 *
 * A request, exactly like taking back your own message: every Nexo client
 * honours it, a modified one need not, and nothing reaches a copy somebody
 * already saved. The dialog that offers it says so.
 */
export async function removeForEveryone(ctx: Context, teamId: string, target: string): Promise<void> {
  await requireModerator(ctx, teamId);
  await conversations.sendPayload(ctx, teamId, { kind: "team_remove", target });
}

/** Renames the team, for everybody in it. */
export function rename(ctx: Context, teamId: string, name: string): Promise<void> {
  const title = name.trim();
  if (title === "") return Promise.reject(new TransportError("rejected", "A team needs a name."));
  return conversations.rename(ctx, teamId, title);
}

/** Says what the team is for. An empty string clears it. */
export async function describe(ctx: Context, teamId: string, description: string): Promise<void> {
  await conversations.sendPayload(ctx, teamId, { kind: "team_meta", description: description.trim() });
}

/**
 * The board, folded from what this device has stored. Only the store: reading
 * what is already here needs no session, no network and no MLS.
 */
export async function board(ctx: Pick<Context, "store">, teamId: string): Promise<Board> {
  const [messages, reactions, marks, team] = await Promise.all([
    ctx.store.messages(teamId),
    ctx.store.reactions(teamId),
    ctx.store.teamMarks(teamId),
    ctx.store.conversation(teamId),
  ]);
  const rows: Parameters<typeof buildBoard>[0] = { messages, reactions, marks };
  if (team?.joinedAt !== undefined) rows.joinedAt = team.joinedAt;
  return buildBoard(rows);
}

async function requireModerator(ctx: Context, teamId: string): Promise<void> {
  let role = await myRole(ctx, teamId);
  if (role === undefined) {
    await roster(ctx, teamId);
    role = await myRole(ctx, teamId);
  }
  if (!moderates(role)) {
    throw new TransportError("rejected", "Only the owner and admins can do that.");
  }
}

File: apps/server/src/teams.rs
//! Teams: an owner, admins and members over an ordinary conversation.
//!
//! A team is a conversation with `kind = 'team'`. Everything its members say
//! to each other -- the name, the description, every post, comment, reaction
//! and pin -- travels as MLS ciphertext through `delivery`, exactly as a group
//! message does, and this module never sees any of it. What it owns is the one
//! thing a team has that a group does not: **who may change who is in it**.
//!
//! That rule lives here, on the server, for the reason `blocks.rs` gives about
//! blocking: a rule only the client applies is one a modified client ignores.
//! The adds and removes themselves stay on `delivery`'s existing routes, which
//! ask [`may_add`] and [`may_remove`] when the conversation is a team -- one
//! membership path, with a gate on it, rather than a second path beside it.
//!
//! What this cannot enforce, and does not pretend to: the inside of an MLS
//! commit. A member running a modified client can build a commit that removes
//! somebody from the group; the server orders commits without reading them.
//! Every group here has that property, and `docs/THREAT-MODEL.md` says so.
//! Likewise pins and admin removals are content, so they are judged by every
//! receiving client against the roles this module serves -- not here.
//!
//! No activity field is served, ever: no "last active", no "online". A roster
//! is who is in the team and what they may do, not when they were last seen.

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::{delete, get, patch, post};
use axum::{Json, Router};
use nexo_protocol::ServerEvent;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::auth::bearer::Caller;
use crate::delivery::DeliveryError;
use crate::state::AppState;

/// The most people a team may hold.
///
/// Commit size and fan-out both grow with the group: every add, removal and
/// key update is a commit every member processes, and every post is an
/// envelope every member's socket is sent. Two hundred keeps that ordinary on a
/// phone. And Nexo is not a broadcast tool -- the public feed is, and reaching
/// thousands of people is what it is for.
pub const MAX_MEMBERS: i64 = 200;

/// What somebody may do in a team.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Role {
    /// Exactly one per team. May do everything, including hand the team on.
    Owner,
    /// May add people, remove members, promote members, pin and remove posts.
    Admin,
    /// May post, comment and react.
    Member,
}

impl Role {
    /// Reads the column. Anything else is refused by the CHECK constraint, so
    /// `None` here means a row this build cannot account for.
    pub fn parse(value: &str) -> Option<Self> {
        match value {
            "owner" => Some(Self::Owner),
            "admin" => Some(Self::Admin),
            "member" => Some(Self::Member),
            _ => None,
        }
    }

    /// The column's spelling.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Owner => "owner",
            Self::Admin => "admin",
            Self::Member => "member",
        }
    }
}

/// Whether someone with `actor`'s role may add people.
pub fn may_add(actor: Role) -> bool {
    matches!(actor, Role::Owner | Role::Admin)
}

/// Whether `actor` may take `target` out of the team.
///
/// Anybody but the owner may take themselves out -- that is leaving. The owner
/// may not, because a team with no owner has nobody who can hand it on or
/// delete it; they transfer first, or delete. Otherwise an admin removes
/// members, and only the owner removes an admin.
pub fn may_remove(actor: Role, target: Role, is_self: bool) -> bool {
    if is_self {
        return actor != Role::Owner;
    }
    match actor {
        Role::Owner => target != Role::Owner,
        Role::Admin => target == Role::Member,
        Role::Member => false,
    }
}

/// Whether `actor` may give `target` the role `to`.
///
/// Only admin and member can be given this way. Ownership moves by
/// [`transfer`], which demotes the old owner in the same transaction; setting
/// it here would make two owners or none. An admin may promote a member; only
/// the owner may demote an admin.
pub fn may_set_role(actor: Role, target: Role, to: Role) -> bool {
    if to == Role::Owner || target == Role::Owner {
        return false;
    }
    match actor {
        Role::Owner => true,
        Role::Admin => target == Role::Member,
        Role::Member => false,
    }
}

/// Every team route. All of them require a [`Caller`].
pub fn router() -> Router<AppState> {
    Router::new()
        .route("/v1/teams", post(create_team).get(list_teams))
        .route("/v1/teams/{id}", delete(delete_team))
        .route("/v1/teams/{id}/members", get(list_members))
        .route("/v1/teams/{id}/members/{handle}", patch(set_role))
        .route("/v1/teams/{id}/transfer", post(transfer))
        .route("/v1/teams/{id}/leave", post(leave))
}

/// The caller's role in a team, or `NotAMember` -- the same answer a
/// conversation that does not exist, or is not a team, gives.
pub(crate) async fn role_in(
    db: impl sqlx::PgExecutor<'_>,
    conversation_id: Uuid,
    user_id: i64,
) -> Result<Role, DeliveryError> {
    let row = sqlx::query!(
        "SELECT m.role FROM conversation_members m
         JOIN conversations c ON c.id = m.conversation_id
         WHERE m.conversation_id = $1 AND m.user_id = $2 AND c.kind = 'team'",
        conversation_id,
        user_id
    )
    .fetch_optional(db)
    .await?
    .ok_or(DeliveryError::NotAMember)?;
    Role::parse(&row.role)
        .ok_or_else(|| DeliveryError::Internal(anyhow::anyhow!("unknown role {:?}", row.role)))
}

/// Holds the team's row for the rest of the transaction.
///
/// Every change to a team's membership or roles takes this first, so two of
/// them on the same team happen one after the other: a transfer and a
/// promotion, two adds racing for the last seat, cannot interleave their reads
/// and writes.
pub(crate) async fn lock_team(
    tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
    conversation_id: Uuid,
) -> Result<(), DeliveryError> {
    sqlx::query!(
        "SELECT id FROM conversations WHERE id = $1 AND kind = 'team' FOR UPDATE",
        conversation_id
    )
    .fetch_optional(&mut **tx)
    .await?
    .ok_or(DeliveryError::NotAMember)?;
    Ok(())
}

/// Tells every member's socket that the roster changed.
///
/// A nudge, not the change -- see [`ServerEvent::Membership`]. Also what ends a
/// removed member's subscription: `stream` re-checks membership when it sees
/// one, rather than on every envelope.
pub(crate) fn announce(state: &AppState, conversation_id: Uuid) {
    state
        .fanout
        .publish(conversation_id, ServerEvent::Membership { conversation_id });
}

#[derive(Deserialize)]
pub struct CreateTeamRequest {
    /// The client picks the id, because it is also the MLS group id and the
    /// group exists on the client before the server hears of it.
    pub conversation_id: Uuid,
}

/// One team, as its member sees it.
#[derive(Serialize)]
pub struct TeamView {
    pub conversation_id: Uuid,
    /// The caller's own role.
    pub role: &'static str,
    pub member_count: i64,
    pub created_at_ms: i64,
}

/// Starts a team with the caller as its owner and only member.
///
/// People are added afterwards through `delivery`'s member route, one commit
/// each, which is where the add gates are. Nothing about the team -- not even
/// its name -- is sent here: the name is the first message.
async fn create_team(
    State(state): State<AppState>,
    caller: Caller,
    Json(request): Json<CreateTeamRequest>,
) -> Result<(StatusCode, Json<TeamView>), DeliveryError> {
    if !state.limits.teams.check(&caller.user_id.to_string()) {
        tracing::warn!(user_id = caller.user_id, "team creation rate limit reached");
        return Err(DeliveryError::TooManyRequests);
    }

    let mut tx = state.db.begin().await?;
    let created = sqlx::query!(
        "INSERT INTO conversations (id, kind) VALUES ($1, 'team')
         ON CONFLICT (id) DO NOTHING
         RETURNING (EXTRACT(EPOCH FROM created_at) * 1000)::BIGINT AS \"created_at_ms!\"",
        request.conversation_id
    )
    .fetch_optional(&mut *tx)
    .await?
    .ok_or_else(|| DeliveryError::Invalid("That conversation already exists.".into()))?;

    sqlx::query!(
        "INSERT INTO conversation_members (conversation_id, user_id, role)
         VALUES ($1, $2, 'owner')",
        request.conversation_id,
        caller.user_id
    )
    .execute(&mut *tx)
    .await?;
    tx.commit().await?;

    Ok((
        StatusCode::CREATED,
        Json(TeamView {
            conversation_id: request.conversation_id,
            role: Role::Owner.as_str(),
            member_count: 1,
            created_at_ms: created.created_at_ms,
        }),
    ))
}

/// The caller's teams.
async fn list_teams(
    State(state): State<AppState>,
    caller: Caller,
) -> Result<Json<Vec<TeamView>>, DeliveryError> {
    let rows = sqlx::query!(
        "SELECT c.id, m.role,
                (SELECT count(*) FROM conversation_members all_m
                 WHERE all_m.conversation_id = c.id) AS \"member_count!\",
                (EXTRACT(EPOCH FROM c.created_at) * 1000)::BIGINT AS \"created_at_ms!\"
         FROM conversations c
         JOIN conversation_members m ON m.conversation_id = c.id
         WHERE m.user_id = $1 AND c.kind = 'team'
         ORDER BY c.created_at",
        caller.user_id
    )
    .fetch_all(&state.db)
    .await?;

    rows.into_iter()
        .map(|row| {
            let role = Role::parse(&row.role).ok_or_else(|| {
                DeliveryError::Internal(anyhow::anyhow!("unknown role {:?}", row.role))
            })?;
            Ok(TeamView {
                conversation_id: row.id,
                role: role.as_str(),
                member_count: row.member_count,
                created_at_ms: row.created_at_ms,
            })
        })
        .collect::<Result<Vec<_>, DeliveryError>>()
        .map(Json)
}

/// One person on a team's roster.
///
/// Three fields, and the absence of a fourth is the design: nothing here says
/// when anybody was last active.
#[derive(Serialize)]
pub struct MemberView {
    pub handle: String,
    pub role: &'static str,
    pub joined_at_ms: i64,
}

/// Who is in a team, owner first, then admins, then members, each by when they
/// joined. For members only.
async fn list_members(
    State(state): State<AppState>,
    caller: Caller,
    Path(conversation_id): Path<Uuid>,
) -> Result<Json<Vec<MemberView>>, DeliveryError> {
    role_in(&state.db, conversation_id, caller.user_id).await?;

    let rows = sqlx::query!(
        "SELECT u.handle, m.role,
                (EXTRACT(EPOCH FROM m.joined_at) * 1000)::BIGINT AS \"joined_at_ms!\"
         FROM conversation_members m
         JOIN users u ON u.id = m.user_id
         WHERE m.conversation_id = $1
         ORDER BY CASE m.role WHEN 'owner' THEN 0 WHEN 'admin' THEN 1 ELSE 2 END,
                  m.joined_at, u.handle",
        conversation_id
    )
    .fetch_all(&state.db)
    .await?;

    rows.into_iter()
        .map(|row| {
            let role = Role::parse(&row.role).ok_or_else(|| {
                DeliveryError::Internal(anyhow::anyhow!("unknown role {:?}", row.role))
            })?;
            Ok(MemberView {
                handle: row.handle,
                role: role.as_str(),
                joined_at_ms: row.joined_at_ms,
            })
        })
        .collect::<Result<Vec<_>, DeliveryError>>()
        .map(Json)
}

#[derive(Deserialize)]
pub struct RoleRequest {
    /// `"admin"` or `"member"`.
    pub role: String,
}

/// Promotes a member to admin, or an admin back to member.
async fn set_role(
    State(state): State<AppState>,
    caller: Caller,
    Path((conversation_id, handle)): Path<(Uuid, String)>,
    Json(request): Json<RoleRequest>,
) -> Result<StatusCode, DeliveryError> {
    if !state.limits.membership.check(&caller.user_id.to_string()) {
        tracing::warn!(user_id = caller.user_id, "membership rate limit reached");
        return Err(DeliveryError::TooManyRequests);
    }
    let to = Role::parse(&request.role)
        .ok_or_else(|| DeliveryError::Invalid("A role is \"admin\" or \"member\".".into()))?;

    let mut tx = state.db.begin().await?;
    lock_team(&mut tx, conversation_id).await?;
    let actor = role_in(&mut *tx, conversation_id, caller.user_id).await?;
    let (target_id, target) = member_by_handle(&mut tx, conversation_id, &handle).await?;

    if !may_set_role(actor, target, to) {
        return Err(DeliveryError::NotPermitted(
            "Only the owner can change an admin, and ownership is handed on, not set.",
        ));
    }
    if target == to {
        return Ok(StatusCode::NO_CONTENT);
    }

    sqlx::query!(
        "UPDATE conversation_members SET role = $3
         WHERE conversation_id = $1 AND user_id = $2",
        conversation_id,
        target_id,
        to.as_str()
    )
    .execute(&mut *tx)
    .await?;
    tx.commit().await?;

    announce(&state, conversation_id);
    Ok(StatusCode::NO_CONTENT)
}

#[derive(Deserialize)]
pub struct TransferRequest {
    /// Who becomes the owner. Must already be in the team.
    pub handle: String,
}

/// Hands the team to somebody else. The old owner stays, as an admin.
///
/// One transaction under the team's lock, demoting before promoting: the
/// one-owner index would refuse the other order, and without the lock two
/// transfers could each read "I am the owner" before either wrote.
async fn transfer(
    State(state): State<AppState>,
    caller: Caller,
    Path(conversation_id): Path<Uuid>,
    Json(request): Json<TransferRequest>,
) -> Result<StatusCode, DeliveryError> {
    if !state.limits.membership.check(&caller.user_id.to_string()) {
        tracing::warn!(user_id = caller.user_id, "membership rate limit reached");
        return Err(DeliveryError::TooManyRequests);
    }

    let mut tx = state.db.begin().await?;
    lock_team(&mut tx, conversation_id).await?;
    if role_in(&mut *tx, conversation_id, caller.user_id).await? != Role::Owner {
        return Err(DeliveryError::NotPermitted(
            "Only the owner can hand the team on.",
        ));
    }
    let (target_id, _) = member_by_handle(&mut tx, conversation_id, &request.handle).await?;
    if target_id == caller.user_id {
        return Err(DeliveryError::Invalid("You already own this team.".into()));
    }

    sqlx::query!(
        "UPDATE conversation_members SET role = 'admin'
         WHERE conversation_id = $1 AND user_id = $2",
        conversation_id,
        caller.user_id
    )
    .execute(&mut *tx)
    .await?;
    sqlx::query!(
        "UPDATE conversation_members SET role = 'owner'
         WHERE conversation_id = $1 AND user_id = $2",
        conversation_id,
        target_id
    )
    .execute(&mut *tx)
    .await?;
    tx.commit().await?;

    announce(&state, conversation_id);
    Ok(StatusCode::NO_CONTENT)
}

/// Takes the caller out of a team. Anybody but the owner.
///
/// Routing only, like every removal here: it stops the caller fetching. The
/// MLS commit that stops them *reading* is made by an owner's or admin's
/// device when it sees the roster change -- a member cannot commit their own
/// removal and keep the rest of the group's keys moving.
async fn leave(
    State(state): State<AppState>,
    caller: Caller,
    Path(conversation_id): Path<Uuid>,
) -> Result<StatusCode, DeliveryError> {
    if !state.limits.membership.check(&caller.user_id.to_string()) {
        tracing::warn!(user_id = caller.user_id, "membership rate limit reached");
        return Err(DeliveryError::TooManyRequests);
    }

    let mut tx = state.db.begin().await?;
    lock_team(&mut tx, conversation_id).await?;
    let role = role_in(&mut *tx, conversation_id, caller.user_id).await?;
    if !may_remove(role, role, true) {
        return Err(DeliveryError::NotPermitted(
            "The owner hands the team on or deletes it, rather than leaving it.",
        ));
    }
    sqlx::query!(
        "DELETE FROM conversation_members WHERE conversation_id = $1 AND user_id = $2",
        conversation_id,
        caller.user_id
    )
    .execute(&mut *tx)
    .await?;
    tx.commit().await?;

    announce(&state, conversation_id);
    Ok(StatusCode::NO_CONTENT)
}

/// Deletes a team for everybody. The owner only.
///
/// The conversation row goes, and its membership and every envelope with it.
/// The members' own copies of what was said are on their devices and are not
/// this server's to reach; the objects its posts' files point at stay in the
/// bucket as ciphertext nobody holds a key for any more.
async fn delete_team(
    State(state): State<AppState>,
    caller: Caller,
    Path(conversation_id): Path<Uuid>,
) -> Result<StatusCode, DeliveryError> {
    if !state.limits.membership.check(&caller.user_id.to_string()) {
        tracing::warn!(user_id = caller.user_id, "membership rate limit reached");
        return Err(DeliveryError::TooManyRequests);
    }

    let mut tx = state.db.begin().await?;
    lock_team(&mut tx, conversation_id).await?;
    if role_in(&mut *tx, conversation_id, caller.user_id).await? != Role::Owner {
        return Err(DeliveryError::NotPermitted(
            "Only the owner can delete the team.",
        ));
    }
    sqlx::query!("DELETE FROM conversations WHERE id = $1", conversation_id)
        .execute(&mut *tx)
        .await?;
    tx.commit().await?;

    announce(&state, conversation_id);
    Ok(StatusCode::NO_CONTENT)
}

/// A member of this team, by handle, with their role.
async fn member_by_handle(
    tx: &mut sqlx::Transaction<'_, sqlx::Postgres>,
    conversation_id: Uuid,
    handle: &str,
) -> Result<(i64, Role), DeliveryError> {
    let row = sqlx::query!(
        "SELECT u.id, m.role FROM users u
         JOIN conversation_members m ON m.user_id = u.id
         WHERE m.conversation_id = $1 AND u.handle = $2",
        conversation_id,
        handle as _
    )
    .fetch_optional(&mut **tx)
    .await?
    .ok_or(DeliveryError::NotFound("member"))?;
    let role = Role::parse(&row.role)
        .ok_or_else(|| DeliveryError::Internal(anyhow::anyhow!("unknown role {:?}", row.role)))?;
    Ok((row.id, role))
}

#[cfg(test)]
mod tests {
    use super::*;
    use Role::{Admin, Member, Owner};

    #[test]
    fn only_owners_and_admins_add() {
        assert!(may_add(Owner));
        assert!(may_add(Admin));
        assert!(!may_add(Member));
    }

    #[test]
    fn the_owner_cannot_leave_and_everyone_else_can() {
        assert!(!may_remove(Owner, Owner, true));
        assert!(may_remove(Admin, Admin, true));
        assert!(may_remove(Member, Member, true));
    }

    #[test]
    fn removal_goes_down_the_ladder_only() {
        assert!(may_remove(Owner, Admin, false));
        assert!(may_remove(Owner, Member, false));
        assert!(may_remove(Admin, Member, false));
        // An admin cannot remove the owner or another admin; only the owner
        // removes an admin.
        assert!(!may_remove(Admin, Owner, false));
        assert!(!may_remove(Admin, Admin, false));
        assert!(!may_remove(Member, Member, false));
        assert!(!may_remove(Member, Admin, false));
    }

    #[test]
    fn roles_move_between_admin_and_member_and_never_to_owner() {
        assert!(may_set_role(Owner, Member, Admin));
        assert!(may_set_role(Owner, Admin, Member));
        assert!(may_set_role(Admin, Member, Admin));
        // Demoting an admin is the owner's alone.
        assert!(!may_set_role(Admin, Admin, Member));
        assert!(!may_set_role(Member, Member, Admin));
        // Ownership is transferred, not set, and the owner is not demoted here.
        assert!(!may_set_role(Owner, Admin, Owner));
        assert!(!may_set_role(Owner, Owner, Admin));
    }

    #[test]
    fn roles_read_back_as_written() {
        for role in [Owner, Admin, Member] {
            assert_eq!(Role::parse(role.as_str()), Some(role));
        }
        assert_eq!(Role::parse("editor"), None);
    }
}
